proxyrunReq_test.go

```diff
@@ -0,0 +1,229 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type mockResponseWriterWithCode struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+// TestRunQueryFail function statResponseWriter's statusCode will not be 200, but the query has been proxied
+// The request will be retried 1 time with a different host in the same replica
+func TestRunQueryFail(t *testing.T) {
+	req := newRequest("http://localhost:8080")
+
+	hs := []string{"localhost:8080", "localhost:8081"}
+
+	s := newMockScope(hs)
+
+	srw := mockStatResponseWriter(s)
+
+	mrw := &mockResponseWriterWithCode{
+		statusCode: 0,
+	}
+
+	retryNum := 1
+
+	_, err := executeWithRetry(
+		context.Background(),
+		s,
+		retryNum,
+		mockReverseProxy,
+		mrw,
+		srw,
+		req,
+		func(f float64) {},
+	)
+
+	if srw.statusCode == 200 {
+		t.Errorf("the retry should be failed: %v", err)
+	}
+
+	if err != nil {
+		t.Errorf("the query should be proxied")
+	}
+}
+
+// TestRunQuerySuccessOnce function statResponseWriter's statusCode will be StatusOK after executeWithRetry, the query has been proxied
+// The execution will succeeded without retry
+func TestRunQuerySuccessOnce(t *testing.T) {
+	req := newRequest("http://localhost:8090")
+
+	hs := []string{"localhost:8080", "localhost:8090"}
+
+	s := newMockScope(hs)
+
+	srw := mockStatResponseWriter(s)
+
+	mrw := &mockResponseWriterWithCode{
+		statusCode: 0,
+	}
+
+	retryNum := 1
+
+	_, err := executeWithRetry(
+		context.Background(),
+		s,
+		retryNum,
+		mockReverseProxy,
+		mrw,
+		srw,
+		req,
+		func(f float64) {},
+	)
+	if srw.statusCode != 200 {
+		t.Errorf("the retry is failed: %v", err)
+	}
+}
+
+// TestRunQuerySuccess function statResponseWriter's statusCode will be StatusOK after executeWithRetry, the query has been proxied
+// The execution will succeeded after retry
+func TestRunQuerySuccess(t *testing.T) {
+	req := newRequest("http://localhost:8080")
+
+	hs := []string{"localhost:8080", "localhost:8090"}
```

Comment:
Can you make sure that both hosts were queried? I don't see such assert in the test. You should do it also in the previous two tests

Reply:
Done